internal/pkg/agent/application/actions/handlers/handler_action_upgrade_test.go

```diff
@@ -368,6 +376,144 @@ func TestUpgradeHandlerNewVersion(t *testing.T) {
+					SourceURI: "http://localhost",
+				},
+			},
+			shouldProxyToEndpoint: true,
+		},
+		{
+			name: "error from coordinator upgrade without notify endpoint",
+			upgradeAction: &fleetapi.ActionUpgrade{
+				ActionType: fleetapi.ActionTypeUpgrade,
+				Data: fleetapi.ActionUpgradeData{
+					Version:   "255.0.0",
+					SourceURI: "http://localhost",
+				},
+			},
+			coordUpgradeErr: errors.New("test error"),
+		},
+		{
+			name: "no error from coordinator upgrade without notify endpoint",
+			upgradeAction: &fleetapi.ActionUpgrade{
+				ActionType: fleetapi.ActionTypeUpgrade,
+				Data: fleetapi.ActionUpgradeData{
+					Version:   "255.0.0",
+					SourceURI: "http://localhost",
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCoordinator := mockhandlers.NewUpgradeCoordinator(t)
+
+			var coordState coordinator.State
+			if tc.shouldProxyToEndpoint {
+				coordState.Components = []runtime.ComponentComponentState{
+					{
+						Component: component.Component{
+							InputSpec: &component.InputRuntimeSpec{
+								Spec: component.InputSpec{
+									ProxiedActions: []string{fleetapi.ActionTypeUpgrade},
+								},
+							},
+							InputType: "endpoint",
+							Units: []component.Unit{
+								{
+									Type: client.UnitTypeInput,
+									Config: &proto.UnitExpectedConfig{
+										Type: "endpoint",
+									},
+								},
+							},
+						},
+					},
+				}
+
+				mockCoordinator.EXPECT().State().Return(coordState)
+			}
+
+			upgradeCalledChan := make(chan struct{})
+			mockCoordinator.EXPECT().Upgrade(mock.Anything, tc.upgradeAction.Data.Version, tc.upgradeAction.Data.SourceURI, mock.Anything, mock.Anything).
+				RunAndReturn(func(ctx context.Context, s string, s2 string, actionUpgrade *fleetapi.ActionUpgrade, opt ...coordinator.UpgradeOpt) error {
+					upgradeCalledChan <- struct{}{}
+					return tc.coordUpgradeErr
+				})
+
+			log, _ := logger.New("", false)
+			u := NewUpgrade(log, mockCoordinator)
+			u.tamperProtectionFn = func() bool { return tc.shouldProxyToEndpoint }
+
+			notifyUnitsCalled := atomic.Bool{}
+			u.notifyUnitsOfProxiedAction = func(ctx context.Context, log *logp.Logger, action dispatchableAction, ucs []unitWithComponent, performAction performActionFunc) error {
+				notifyUnitsCalled.Store(true)
+				return nil
+			}
+
+			ack := mockAcker.NewAcker(t)
+
+			if tc.coordUpgradeErr != nil {
+				// on a coordinator upgrade error we should ack and commit all the bkg actions
+				ack.EXPECT().Ack(mock.Anything, mock.Anything).Return(nil)
+				ack.EXPECT().Commit(mock.Anything).Return(nil)
+			}
+
+			err := u.Handle(ctx, tc.upgradeAction, ack)
+			require.NoError(t, err, "Handle should not return an error")
+
+			select {
+			case <-upgradeCalledChan:
+				break
+			case <-time.After(10 * time.Second):
+				t.Fatal("mockCoordinator.Upgrade was not called in time")
+			}
+
+			assert.False(t, notifyUnitsCalled.Load(), "notifyUnitsOfProxiedAction should not be called")
```

Comment:
It's not obvious why this is unconditionally false, instead of being dependent on the value of `tc.shouldProxyToEndpoint`.

I also would expect a test where this function is called but returns an error to handle that case properly (which is also what I thought the point of this fix was but perhaps I am missing something fundamental).

Reply:
Thanks for the note! Yeah, this is always expected to be false - the idea is to validate that the func is never called directly, but only passed as a `PreUpgradeCallback` to the coordinator upgrader. It’s mocked anyway, so even if I had a test where this exact callback returns an error, it wouldn’t add much value.

That said, I did add a dedicated unit test on the coordinator side with a callback that returns an error, and there I verify both that the error is propagated and that the upgrade details are marked as failed:
https://github.com/elastic/elastic-agent/pull/9634/files#diff-21de8fb82a5ebccaa0ca3afb0cb8bfbe10d5b8f920021785c2d485ee35a3556cR544-R586

IMO what you’re suggesting would essentially be equivalent to that test. Wdyt?

Reply:
OK yes I see now I was reading this test expecting it to do what the other one did, which was what was misleading me.

This line is verifying that `notifyUnitsOfProxiedAction` is never called directly as you say which is a great check to had. It would be clearer if you added this context either directly to the error message in the assertion or as a comment in the code.

Reply:
I can now see that the two tests together verify the properties we need:
1. This file verifying that errors from the coordinator upgrade handler result in background actions being updated appropriately
2. The coordinator unit test verifying that a pre upgrade check returning an error returns a coordinator error.

Reply:
added an explanation for this assertion here 9c2f633555d38472cee41e0251d87ef190366b47